docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,303 @@
+
+```
+-- Convert a STRING to the 2.0 INET type.
+-- This command will return a row containing a job id.
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  DEFERRED
+
+-- The application can continue reading and writing this column as STRING.
+
+-- Wait for the job to finish, stop the application, swap the column type.
+-- This command would block, returning any output from the conversion job.
+-- Any reads that happen-after this command will see the new type, while any
+-- reads that happen-before this command will see the old type.
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  COMPLETE
+
+-- Restart new version of the application that expects the INET type.
+```
+
+Electing not to use the `DEFERRED` option is equivalent to running a
+`DEFERRED` followed immediately by a `COMPLETE`.  We will refer to this
+as "auto-complete" mode.
+
+The `COMPLETE` option requires the entire, original `ALTER TABLE`
+command as a sanity-check.  If the type conversion generated any output,
+it will be emitted by this command.  If the type conversion has not yet
+completed, this command will block until it has.  It is not an error to
+run `COMPLETE` before the conversion job has finished; it is simply a
+signal to indicate that the SQL consumers are prepared to handle the new
+column type.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC.  It could be possible to support
+foreign keys in the future by cascading the type change, but this would
+greatly increase the amount of coordination required to effect the total
+schema change.
+
+# Reference-level explanation
+
+## Process
+
+Conceptually, a column type change can be thought of as creating a
+shadow column of the new type, which captures writes to the original,
+source, column while a backfill process is run.  Once the backfill is
+complete, the shadow column is swapped for the original, which is then
+retired.  The shadow column will be anonymous and generally invisible to
+the user until it is swapped in.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER TABLE ... COLUMN ... TYPE` is run, resulting in a new shadow
+  column being created in `DELETE_ONLY` mode.  The shadow column has a
+  new column id and a reference to the source column
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
+* The shadow column and indexes are upgraded to `WRITE_ONLY` mode.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column
+  and indexes.
+* A backfill process is distributed across all nodes in the cluster
+  which have data for the table.  This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
```

Comment:
I don't think we can insert data with the source MVCC timestamp. If the timestamp is below the GC threshold then we certainly can't commit a txn at that timestamp. Also, that would require one txn per row, which is really slow. I think it's ok to use whatever the backfill chunk read time is, just like for column backfill.